### PR TITLE
feat: Controlled Cells

### PR DESCRIPTION
**Pull request by lonnieezell**
Expands Cell functionality by bringing a new Cell class and formal functionality. The Cell class can help quickly build a cell, and provides several handy capabilities for making views cleaner and more flexible to work with.

### CODE REVIEW

**Review by michalsn — COMMENTED** [review]
I pretty much love it. Feels like a solid structure. When we start using it, maybe there will be a place for improvements (as always), but this is great as it is.

I wasn't sure about the `view` handling, but it seems like a good compromise in exchange for handy features. 

I also like the computed properties feature that gives us access to the private/protected variables.

I only have one main question regarding Controlled Cells - shouldn't we force the use of the `render` method? For me, components should have only one point of entry when we want to render them. So leaving the possibility to call another method from `view_cell()` seems not right. I'm saying this mainly because of the planned improvements - Live Cells.

Also, I think you should run CS Fixer on this code. I spotted many small things, but I'm pretty sure they can be fixed automatically, so I didn't bother to point them out.

**Comment by lonnieezell** [comment]
> I only have one main question regarding Controlled Cells - shouldn't we force the use of the render method? For me, components should have only one point of entry when we want to render them. So leaving the possibility to call another method from view_cell() seems not right. I'm saying this mainly because of the planned improvements - Live Cells.

@michalsn I was on the fence about that. Until we have live cells it might be handy to call other methods. I know you said you called one method on first output, and another on updates currently. If I had to guess on timing, I would think controlled cells would come out in 4.3 and live cells would take a bit longer to get in a good place, so likely the next major release. Should we force that now? 

My planned roadmap for the very short future is: 

- cell & view generator command 
- likely the $routes->cell() method I'd mentioned on the forums.  Part of me thinks we might be able to get a different version of Live Cells available relatively easily if the `route->cells()` method could accept a "control array" that contained things like: 

```php
   callMethod => [
      'name' => 'methodFoo', 
      'params': ['foo' => 'bar]
   ],
   setValue: [
      'key' => 'foo', 
      'value' => 'baz'
   ]
]
```

That would allow us to keep using different HTTP methods, but then have the POST/PUT methods able to support this type of command array, that could then call a `hydrate()` method (or some other better named method) on the Cell. 

That will take experimentation and a bit of a think about security issues with that. 

Basically, I guess I'm saying I'm concerned about locking us in too far currently since the future is a bit unknown. I suppose we could always have a `live_cell` helper that would handle it in a more restricted way. We also might have more life-cycle events to deal with when we get there.

**Comment by michalsn** [comment]
>  I know you said you called one method on first output, and another on updates currently.

Fair point, but that's because my lazy ass didn't pay attention to the `view_cell` description in the user guide... somehow I figured out that the only way to pass parameters to a class is through an array. And that's why I ended up with two methods 😅 

Ok, question. If I call `$routes->cell('counter/increment')` - are you going to call only the `increment()` method or also the `render()` method? This is important because maybe next to the `live_cell` helper, we will need the `$routes->live_cell()` method as well.

> That would allow us to keep using different HTTP methods, but then have the POST/PUT methods able to support this type of command array, that could then call a hydrate() method (or some other better named method) on the Cell.

As for passing parameters to a Cell from a request, I have no opinion so far, and I guess I need to do some research/reading on this topic. If we care about using different HTTP verbs, then the answer is simple. However, my first feelings were to send everything via POST as JSON. Because that would simply be the easiest solution. Does it somehow make a difference for the developer? I don't think so, but by accident, it would increase security (CSRF). We could also consider some sort of HTTP method spoofing - so we could still use different HTTP verbs in our routes. I'm not sure, really - I'm making this up as I write.

In any case, yes - you're right. We shouldn't lock ourselves at this stage. There are still a lot of unknowns ahead but the roadmap seems great.

**Comment by lonnieezell** [comment]
> Ok, question. If I call $routes->cell('counter/increment') - are you going to call only the increment() method or also the render() method? This is important because maybe next to the live_cell helper, we will need the $routes->live_cell() method as well.

Render method would be called every time. If we handle it like LiveWire, there's a single controller that handles all of it. How would work best for us? Not sure, honestly. 

> In any case, yes - you're right. We shouldn't lock ourselves at this stage. There are still a lot of unknowns ahead but the roadmap seems great.

Cool. I'll get the rest of the things cleaned up now and the PR updated.

**Comment by michalsn** [comment]
> Render method would be called every time.

Great.

> If we handle it like LiveWire, there's a single controller that handles all of it. How would work best for us? Not sure, honestly.

TBH I was holding myself back from digging into LIveWire (and others) so far. I prefer to start with a fresh mind - without being directed to something. Not that I think I can make any architectural breakthrough on this matter, but still, I am trying to keep my mind open.

We don't necessarily have to have one controller. They probably did it this way to simplify a couple of things on their side. I guess it's about time for me to see how LiveWire works.

We won't get away from experimenting to choose the best path to fit CodeIgniter's philosophy... Luckily, with this PR being merged, we will have an excellent place for doing those experiments.

**Review by michalsn — APPROVED** [review]
Coveralls amazes me sometimes... there is an increase everywhere, but in summary, we have a decrease 😄

**Comment by lonnieezell** [comment]
> TBH I was holding myself back from digging into LIveWire (and others) so far. I prefer to start with a fresh mind - without being directed to something. Not that I think I can make any architectural breakthrough on this matter, but still, I am trying to keep my mind open.

That's a good idea. And will be a definite help with brainstorming and experimenting. 

> We don't necessarily have to have one controller. They probably did it this way to simplify a couple of things on their side. I guess it's about time for me to see how LiveWire works.

I agree - I think it's probably we could get away with allowing multiple controllers. Definitely looking forward to experimenting some more in that area, though. Finding the path of least resistance for both the developer and the framework.

File: user_guide_src/source/outgoing/view_cells.rst [hunk 1]
```diff
@@ -56,6 +64,176 @@ When you use it this way, all of the parameters must always be specified in the
         return view('recentPosts', ['posts' => $posts]);
     }
 
+.. _controlled-cells:
+
+****************
+Controlled Cells
+****************
+
+Controlled Cells have two primary goals: to make it as fast as possible to build the cell, and provide additional logic and flexibility to your views, if they need it. The class must extend ``CodeIgniter\View\Cells\Cell``. They should have a view file in the same folder. By convention the class name should be PascalCase and the view should be the snake_cased version of the class name. So, for example, if you have a ``MyCell`` class, the view file should be ``my_cell.php``.
```
[the remark this answers is not in the document]
Comment (lonnieezell) [reply user_guide_src/source/outgoing/view_cells.rst:73]: I hadn't noticed that since on non-case-sensitive file-systems they do display together. At least it does on my Mac, and I assume Windows would do the same thing. Looking at GitHub's file listings and it doesn't, though, which I assume is it running on Linux. 

I'm not sure the best solution here. By convention, views in CI have always been lowercase, which is why I went that route. They've also never needed special extensions other than `.php`. 

We could pull a book out of Next and SvelteKit and require them to be in their own folder, but that seems like unnecessary complexity. 

While it would be ideal if it was next to each other in all situations if I'm being honest, I'm not too concerned. In typical MVC setups they're not next to each other either, Controllers and Views that is. I realize it's not a perfect comparison, too....

[the remark this answers is not in the document]
Comment (lonnieezell) [reply user_guide_src/source/outgoing/view_cells.rst:73]: I was working on the generator command for Cells tonight and it seems the generator command will automatically force uppercase on the file name. Which gave me the happy accident of having the following files: 

```
- Test.php
- TestCell.php
```

While that goes against our current conventions with views being lower-cased, it's not a terrible thing to force the class to have the Cell postfix and the view be the name without. Though I'll be honest, it took me a moment to realize that `Test.php` was the view and not an artifact leftover from a previous run. :)

[the remark this answers is not in the document]
Comment (datamweb) [reply user_guide_src/source/outgoing/view_cells.rst:73]: I remember, the generator converts names like `name-name-tow` to `NameNametow`.

[the remark this answers is not in the document]
Comment (paulbalandan) [reply user_guide_src/source/outgoing/view_cells.rst:73]: > ... it seems the generator command will automatically force uppercase on the file name. 

When supplying the `name` argument to the generator command call, it is assumed that the `name` will be used for a class file. That's why the class name and file name are turned to PascalCase.

[the remark this answers is not in the document]
Comment (michalsn) [reply user_guide_src/source/outgoing/view_cells.rst:73]: From what I see, the only official package that uses view files (Shield) uses a `sneak_case` convention. So in my opinion, it would be a good idea to continue this.

Although I'm not going to hide the fact that I've also always used this convention for views.

@paulbalandan Are there any chances to easily implement the configuration of this from the command class level? Could we add something like `depascalize` for the class name only?

```
function depascalize(string $string): string
{
    return strtolower(substr(preg_replace('/([A-Z])/', '_$1', $string), 1));
}
```

[the remark this answers is not in the document]
Comment (paulbalandan) [reply user_guide_src/source/outgoing/view_cells.rst:73]: If I will be making a `make:cell` command, I would add a `getView()` method where I would "depascalize" the file path. My assumption is this `make:cell` command will produce two files, the custom `Cell` and its associated view file made thru `getView()`.

**Comment by MGatner** [comment]
I've snoozed this one for a few days until I had more time - grateful for @michalsn's review! I will still need to go through the code before I understand it all but I've read the whole conversation and it sounds like this is going on a great direction. Props to @lonnieezell for breaking up live cells into mergable bites!
